2단계 - 연관 관계 매핑

seogineer opened the pull request:
PR 올립니다.

seogineer committed:
fk_answer_to_question

seogineer committed:
feat: Answer - User OneToOne

seogineer committed:
feat: Question - User OneToOne

seogineer committed:
feat: DeleteHistory - User OneToOne

seogineer committed:
test: 테스트 데이터 추가

ohtaeg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -18,9 +20,11 @@ public class Answer extends BaseTimeEntity {
     private Long id;
 
-    private Long writerId;
+    @OneToOne
+    private User writer;
```
User와 1:1 관계를 맺어주셨군요 :) 👍 

수정을 위한 리뷰가 아니오니 참고만 해주세요!

개인적으로 User와 연관 관계를 맺어도 좋지만 꼭 필요한 경우가 아니면 간접 참조를 우선으로 고려합니다 ㅎㅎ
writer_id 로만 활용해도 충분하다고 생각이 되어지기도 해서 코멘트 드려봅니다 :)

이에 대해 도경님의 의견도 궁금하네요 😄

seogineer replied:
미션이 연관 관계 매핑이라서 과하게 관계 설정을 한 듯 합니다.

seogineer replied:
미션에서 힌트로 제공한 쿼리인데 아래 맞추다보니 이렇게 만들게 된것 같습니다.
OneToOne은 경험이 없어서 잘 한건지 모르겠습니다....

alter table answer
    add constraint fk_answer_to_question
        foreign key (question_id)
            references question

alter table answer
    add constraint fk_answer_writer
        foreign key (writer_id)
            references user

alter table delete_history
    add constraint fk_delete_history_to_user
        foreign key (deleted_by_id)
            references user

alter table question
    add constraint fk_question_writer
        foreign key (writer_id)
            references user

seogineer replied:
일단 짚어주신 코드는 간접 참조로 수정했는데 나머지 OneToOne도 간접 참조가 우선이니 수정하는게 좋을까요?

ohtaeg replied:
앗 꼭 변경해주실 필요는 없어요!! 잘해주셨어요 👍 
위 피드백에 말씀드렸듯이 수정을 위한 리뷰는 아니였습니다 :)
간접참조에 대해서도 한번 고민해보자 라는 취지의 피드백이였습니다 😄 
이번 미션은 JPA 학습에 초점에 맞춰져 있기 때문에 도경님 마음대로 간접참조로 변경해주셔도 좋고 유지해주셔도 좋습니다!

ohtaeg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -18,9 +20,11 @@ public class Answer extends BaseTimeEntity {
+    private User writer;
 
-    private Long questionId;
+    @ManyToOne
```
ManyToOne 의 기본 FetchType 은 EAGER 인데요!
즉시 로딩으로 해주신 의도가 있으신가요?
즉시 로딩으로 하게 되면 어떤 문제점이 있을까요 :) ?

seogineer replied:
left join으로 jpql이 생성되기 때문에 questionId가 null이어도 Answer가 조회되는 상황이 발생할것 같습니다.

ohtaeg commented on src/main/java/qna/domain/AnswerRepository.java:
```diff
@@ -7,7 +7,7 @@
 
 @Repository
 public interface AnswerRepository extends JpaRepository<Answer, Long> {
```
현재 도메인 패키지에 repository가 있는데 도메인 패키지와 분리해주시면 좋을 것 같아요 :)

ohtaeg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -21,7 +26,11 @@ public class Question extends BaseTimeEntity {
+    @OneToOne
+    private User writer;
+
+    @OneToMany(mappedBy = "question")
```
답변은 질문에 종속되기 때문에 orphanRemoval true 옵션을 줘도 좋을 것 같은데
도경님은 어떻게 생각하시나요 :) ?

ohtaeg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -82,7 +95,8 @@ public String toString() {
                 ", title='" + title + '\'' +
                 ", contents='" + contents + '\'' +
-                ", writerId=" + writerId +
+                ", writer=" + writer +
+                ", answers=" + answers +
```
현재는 즉시 로딩이지만 만약 FetchType.LAZY 로 설정된 대상 객체에서
양방향 연관 관계일 경우 toString 에서 출력할 때 유의해야 할 점은 어떤 게 있을까요 :) ?

seogineer replied:
N + 1 문제가 발생할 수도 있을것 같습니다.

seogineer replied:
question -> answer 방향으로 관계를 제거하겠습니다.

ohtaeg replied:
추가로 서로 계속 호출되는 순환참조도 발생할 수 있을 것 같아요 😄

ohtaeg reviewed (changes requested):
2단계도 잘해주셨습니다 👍 👍 
마지막 미션을 진행하기전에 한번 고민해보면 좋을 포인트가 존재하여 Request changes 하였습니다!
소소한 코멘트를 드렸는데 확인해주시면 감사하겠습니다 :D

seogineer committed:
style: 클래스 용도별 패키지 분리

seogineer committed:
refactor: 관계 매핑 수정

- User writer -> Long writerId
- Question @ManyToOne -> @ManyToOne(fetch = FetchType.LAZY)

seogineer committed:
style: 불필요한 import 제거

seogineer committed:
refactor: Questiona-Answer orphanRemoval 추가

ohtaeg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -29,7 +31,12 @@ public class Question extends BaseTimeEntity {
+    @OneToMany(
+        mappedBy = "question",
+        fetch = FetchType.LAZY,
+        cascade = CascadeType.PERSIST,
```
여기에서의 cascade 설정은 어떤 동작을 예상하고 사용하셨을까요 :) ?

[CascadeType.PERSIST 사용을 조심히 해야하는 이유](https://joont92.github.io/jpa/CascadeType-PERSIST%EB%A5%BC-%ED%95%A8%EB%B6%80%EB%A1%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%A9%B4-%EC%95%88%EB%90%98%EB%8A%94-%EC%9D%B4%EC%9C%A0/)

seogineer replied:
제가 기능에 대해서 자세히 모르고 사용한것 같습니다. CascadeType.REMOVE를 사용하면 될까요? 아니면 CascadeType 설정은 필요가 없는걸까요?

ohtaeg replied:
위 블로그에서 잘 설명해주신 것처럼 예상치 못한 사이드 이펙트를 발생할 수 있습니다!
이 부분은 제가 알려드리기보다는 곰곰히 생각해보시면 좋을 것 같아요 :)

- CascadeType.PERSIST에 대해 설명을 해볼 수 있으면 좋을 것 같아요!
- 질문 객체가 아직 영속화되지 않았는데 답변을 어떻게 달수 있을까요 :) ?
  - https://bepoz-study-diary.tistory.com/383

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨어요 👍 👍 
소소한 피드백들을 드렸는데 확인해주시면 감사하겠습니다!!
마지막 미션도 화이팅이에요!! 🔥
